emacs_lua_complete.lua

```diff
@@ -0,0 +1,81 @@
+--    __emacs_lua_complete:
+-- This function finds and prints a limited number of matches on the
+-- lua globals tree, _G. See `lua-complete-string`. N.B.: Does not
+-- complete numeric keys (e.g. for arrays) or keys which begin with an
+-- underscore.
+-- Arguments:
+--   PARTS: a table array of dot-separated parts of the completion item
+--     (e.g. table.con would become {"table","con"}).  All but the last
+--     element must be valid subkeys in the _G heirarchy.
+--   LIBS: an array of tables of the form
+--       {var: varname, lib: libfile}
+--     for e.g. local varname=require("libfile"). This library will be
+--     require'd and searched alongside global variables.
+--   LOCALS: an array of strings parsed from local aloc,bloc = ... to
+--     complete as well
+--
+-- Example:
+-- __emacs_lua_complete({'table','ins'},{{var="xyz",lib="libfile"}},
+--                      {"frank","beans"},5000)
+--
+-- Written 2019/05 J.D. Smith
+function __emacs_lua_complete(parts,libs,locals,limit)
+   -- Setup a FIFO queue for breadth-first tree traversal
+   local queue={head=0,tail=0}
+   function queue:push(value)
+      self.tail=self.tail+1  self[self.tail]=value
+   end
+   function queue:pop()
+      self.head=self.head+1  return self[self.head]
+   end
+
+   -- Mirror the globals table
+   local globals,cnt,limit={},0,limit or 2000
+   for k,v in pairs(_G) do globals[k] = v end
+   
+   -- Add libs and locals
+   if libs then
+      for _,v in ipairs(libs) do
+	 local good,lib=pcall(require,v.lib)
+	 globals[v.var]=good and lib or {}
+      end
+   end
+   if locals then
+      for _,v in ipairs(locals) do globals[v]={} end
+   end
+
+   -- Descend the tree to the limit of dotted parts
+   local g,pre=globals,""
+   for i=1,#parts do
+      if not g then break end	-- should always exist
+      if i<#parts then		-- keep descending
+	 pre = (i > 1 and pre.."." or "")..parts[i]
+	 g = g[parts[i]]
+      else  -- final part; seed the queue with any matching tables
+	 for k,v in pairs(g) do
+	    if type(k)=="string" and k:sub(1,1) ~= "_" and k:find('^'..parts[i]) then
+	       local dpath=(i > 1 and pre.."." or "")..k
+	       print(dpath)
+	       cnt=cnt+1
+	       if type(v)=="table" then queue:push({pre=dpath,entries=v}) end
+	    end
+	 end
+      end
+   end
+
+   -- Perform a limited breadth-first traversal of the queued tables
+   while queue.head < queue.tail and cnt < limit do
+      local q=queue:pop()
+      for k,v in pairs(q.entries) do
+	 if type(k) == "string" and k:sub(1,1) ~= "_" then 
+	    pre = q.pre.."."..k
```

Comment:
Remember you told me that when keys come from completions, they are verified that they don't contain whitespaces or dots? Here, if I'm not mistaken, `k` comes from Lua tables and no verification is done, and thus `k` can contain whatever and produce completions that are invalid syntax or don't work as expected. WDYT?

Reply:
Good point.  Yeah any valid key in Lua can be in `k`.  But you can't reference such keys with dot-syntax (I just tried).  Since we are only parsing and completing dot-syntax I think we can get away with it.  It would be better of course to parse `a[" foo "].b[" c"]` style things too, but that's a new can of worms.  If that ever comes to pass, you'd trim and recreate whitespace around dots, but NOT within brackets (since there it is significant and must be preserved).  I guess it might not be that hard to take an arbitrary expression with brackets and dots (and comments) and turn it into a series of parts with all the ignorable whitespace removed and all the relevant whitespace retained.  Right now it's this simple bit:

```lua
 (mapcar 'lua-completion-trim-input (split-string expr "\\."))
```
Wouldn't be terrible to update this to take the expression and remove comments, find brackets etc.  Right now `lua-start-of-expression` doesn't handle brackets gracefully.

Reply:
I think the simplest solution is to just filter out any keys that are not alphanumeric-plus-underscore here.